### PR TITLE
ARROW-12106: [Rust][DataFusion] Support `SELECT * from information_schema.tables`

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
Provide configurable access to a table list so a user can see what tables exist).

See the doc for background: https://docs.google.com/document/d/12cpZUSNPqVH9Z0BBx6O8REu7TFqL-NPPAYCUPpDls1k/edit#

I plan to add support for `SHOW TABLES` (and `information_schema.columns` / `SHOW COLUMNS`) as follow on PRs

# Example use:

Setup:
```
echo "1" > /tmp/table.csv
echo "2" >> /tmp/table.csv
cargo run --bin datafusion-cli
```

Then run :

```
>  CREATE EXTERNAL TABLE t(a int)
STORED AS CSV
LOCATION '/tmp/table.csv';

0 rows in set. Query took 0 seconds.
> select * from information_schema.tables;

+---------------+--------------------+------------+--------------+
| table_catalog | table_schema       | table_name | table_type   |
+---------------+--------------------+------------+--------------+
| datafusion    | public             | t          | BASE TABLE   |
| datafusion    | information_schema | TABLES     | SYSTEM TABLE |
+---------------+--------------------+------------+--------------+
2 row in set. Query took 0 seconds.
z>
```

### CODE REVIEW

File: rust/datafusion/src/execution/context.rs
Comment (alamb) [inline rust/datafusion/src/execution/context.rs:2080]: These tests illustrate how this feature works concisely I believe

**Comment by alamb** [comment]
FYI @seddonm1  @returnString @Dandandan

File: rust/datafusion/src/catalog/information_schema.rs
Comment (seddonm1) [inline rust/datafusion/src/catalog/information_schema.rs:155]: why 10?

File: rust/datafusion/src/catalog/information_schema.rs
Comment (seddonm1) [inline rust/datafusion/src/catalog/information_schema.rs:176]: I have not really spent much time with the information_schema. Are the own two types `BASE TABLE` and `SYSTEM TABLE`?

**Comment by seddonm1** [comment]
@alamb my read through looked good (as expected). I can have a play with it early next week.

File: rust/datafusion/src/catalog/information_schema.rs
Comment (alamb) [reply rust/datafusion/src/catalog/information_schema.rs:176]: I am not sure to be honest. Postgres has `BASE TABLE` and `VIEW` that I saw.

I will spend some time figuring out how what the values of these should be.

Comment (seddonm1) [reply rust/datafusion/src/catalog/information_schema.rs:176]: Ok. Maybe we can find the standards document somewhere as I assume there is a fixed list.

**Comment by returnString** [comment]
This looks awesome (and so hot on the heels of the basic catalog impl!) - will take a look tomorrow :D

File: rust/datafusion/src/catalog/catalog.rs
Comment (Dandandan) [inline rust/datafusion/src/catalog/catalog.rs:41]: Better as `&[&str]` to avoid clone?

Comment (Dandandan) [reply rust/datafusion/src/catalog/catalog.rs:41]: ah I see in the implementation it is not really possible

File: rust/datafusion/src/execution/context.rs
Comment (Dandandan) [inline rust/datafusion/src/execution/context.rs:668]: Can it be disabled as well - or don't we expect that?

File: rust/datafusion/src/catalog/information_schema.rs
Comment (seddonm1) [reply rust/datafusion/src/catalog/information_schema.rs:176]: @alamb Here is a link to the Postgres doc: https://www.postgresql.org/docs/current/infoschema-tables.html

Type of the table: 
- `BASE TABLE` for a persistent base table (the normal table type)
- `VIEW` for a view
- `FOREIGN` for a foreign table 
- `LOCAL TEMPORARY` for a temporary table

Are all our 'tables' kind of `LOCAL TEMPORARY`?

File: rust/datafusion/src/execution/context.rs
Comment (returnString) [reply rust/datafusion/src/execution/context.rs:668]: Would it be worth enabling by default, even, and adding an option to _disable_ it here? I can't think of any downsides to enabling it, and for the subset of the info schema that we support, it's in line with the standard (as far as I'm aware without spending stupid money of a copy of the standard 😅)

File: rust/datafusion/src/execution/context.rs
Comment (returnString) [inline rust/datafusion/src/execution/context.rs:703]: Not a massive issue as this isn't the critical path, but would it be worth doing this wrapping on catalog registration so we avoid query-side allocs here? I wonder if it's possible to remove the `CatalogListWithInformationSchema` entirely and just do this inside `register_catalog` instead. So if the info schema is enabled, wrap the catalog up in a `CatalogWithInformationSchema` before registering it with the list, or otherwise, register the plain catalog.

**Review by returnString — CHANGES_REQUESTED** [review]

File: rust/datafusion/src/catalog/information_schema.rs
Comment (alamb) [reply rust/datafusion/src/catalog/information_schema.rs:176]: I think `LOCAL TEMPORARY` is referring to what happens when you do `CREATE TEMPORARY TABLE...` in postgres -- so I agree with @returnString  that LOCAL TEMPORARY is probably not what we would want.

I will change the information table to say `VIEW` for the information_schema tables, consistent with Postgres (and SQL server)

File: rust/datafusion/src/catalog/catalog.rs
Comment (alamb) [reply rust/datafusion/src/catalog/catalog.rs:41]: The way I could think of to avoid the copy is don't make a `CatalogList` trait and instead pass around `MemoryCatalogList` directly. This would allow a function like `fn catalog_names(&self) -> impl IntoIterator<Item=&str>` 

However, given `catalog_names` are only called when creating the information_schema views, I am not sure this optimization is warranted at this time

File: rust/datafusion/src/execution/context.rs
Comment (alamb) [reply rust/datafusion/src/execution/context.rs:668]: > Can it be disabled as well - or don't we expect that?

I didn't expect people to want to be able to change the setting at runtime on the same `Context` -- either they would want the information schema or not. However, for consistency with the other settings I think it would be good to allow the flag to be passed in and I will do so

> Would it be worth enabling by default, even, and adding an option to disable it here? 

There might be some small runtime overhead for always enabling it. I don't have any strong preference on one way or the other. I am happy to change the default. Does anyone else have any thoughts?

**Comment by alamb** [comment]
> I could definitely see an argument that the csv/parquet TableProviders are temporary, but arbitrary TableProviders might use totally different semantics (e.g. in my case, fetching data from a persistent store that could be classed as a base table). If we want to support different options here, perhaps we need a new TableProvider method to indicate that?

Yes, I agree -- I see no reason to support different values for `TABLE_TYPE` in the initial PR -- and if someone wants that functionality they can perhaps propose a PR to add a methods to the `TableProvider`

File: rust/datafusion/src/execution/context.rs
Comment (alamb) [reply rust/datafusion/src/execution/context.rs:703]: Since the `Config` can't be changed after the `ExecutionContext` is created I think this would be a good idea. I will give it a try.

File: rust/datafusion/src/catalog/information_schema.rs
Comment (alamb) [reply rust/datafusion/src/catalog/information_schema.rs:155]: The string builder requires a number here but the number or rows isn't known. I'll add a comment explaining 10 is some arbitrary number

Comment (Dandandan) [reply rust/datafusion/src/catalog/information_schema.rs:155]: Why not use `0` to keep it simple?

Comment (alamb) [reply rust/datafusion/src/catalog/information_schema.rs]: I guess I felt that `10` was a siimple as `0` -- lol though I don't really care

File: rust/datafusion/src/execution/context.rs
Comment (alamb) [reply rust/datafusion/src/execution/context.rs]: in https://github.com/apache/arrow/pull/9818/commits/5151f31eecfbcd0d5892837f951bb0e5ea8473c0

**Comment by alamb** [comment]
Ok, I have addressed all comments and I think this PR is ready to merge (once it gets an approval).

File: rust/datafusion/src/execution/context.rs
Comment (returnString) [inline rust/datafusion/src/execution/context.rs:330]: Leftover debugging?

**Review by returnString — APPROVED** [review]
One stray println left in as noted, but otherwise, looks good to me! Looking forward to using this :)

**Comment by returnString** [comment]
I've also logged https://issues.apache.org/jira/browse/ARROW-12132 for the table provider "type" followup.

**Review by jorgecarleitao — COMMENTED** [review]
Hi @alamb . I am sorry: I am obviously missing something: I do not understand the rational of introducing catalogs and schemas.

In the proposal in google docs, the benefits enumerated are the ability to introspect the table and its schemas (i.e. column and types), which I very much agree with. I am just trying to understand the rational for catalogs and schemas (grouping of table names).

IMO Catalogs in postgres are a mechanism over which postgres uses its own system (RDBMS) [to perform bookeeping about its own state](https://www.postgresql.org/docs/9.1/catalogs.html), away from normal users.

IMO Schemas in postgres are primarily a mechanism over which postgres adds an extra layer of granularity over groups of tables, typically in the context of its role-based model and permission segmentation. Since DataFusion has no concept of user or role or permission, I do not see the why.

There is another reason in postgres to use schemas, and that may be relevant here: schemas reduce naming collisions by allowing users to defining tables in their own schema. However, I only see this being relevant whenever an administrator can control which users can write to which schemas, as that induces a "namespace cleanliness" of the schema (i.e. the DBA can promise the user that the schema is "clean"). IMO such promise requires a role-based model and a permission system attached, no?

Maybe the reason we use catalogs and schemas here is to mimic postgres behavior so that consumers can more easily query using the same SQL?

Sorry for the long comment, I am just trying to understand the logic here.

**Comment by returnString** [comment]
Hey @jorgecarleitao, so I introduced the general concept in #9762 (design doc [here](https://docs.google.com/document/d/1_bCP_tjVRLJyOrMBOezSFNpF0hwPa1ZS_qMWv1uvtS4/edit#)), with a mix of rationales

Firstly, I saw Andrew's proposal for this info schema work and wanted to support namespacing this correctly, as it works in other DBs (we don't want to disallow people from registering a table called `tables`, for the sake of a silly example, and having that shadow the system one).

Secondly, in my use case (DataFusion as the execution engine for an actual DB), I have catalogs and schemas as a mechanism for logical organisation. From the proposal doc:
> - Multi-tenant software seeking more stringent data isolation might use catalogs to represent individual tenants
> - Schemas could be used to split tables by producer in a centralised data pipeline
> - In a data warehouse comprising both raw event logs and pre-aggregated summary tables, you could use “raw” and “summary” schemas to organise the tables by type

Your point about roles/permissions is totally correct - in my case currently, I'm dynamically building execution contexts based on tables that users are allowed to access according to my external metadata. Users can also create tables, which is again a metadata operation that informs future execution context creation (and these are created within a specific catalog and schema).

Lemme know if anything there is unclear, doing lunchbreak replies atm so may have missed something 😄

File: rust/datafusion/src/execution/context.rs
Comment (alamb) [reply rust/datafusion/src/execution/context.rs:330]: whoops -- good call -- will remove.

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Thanks a lot @alamb and @returnString for this. Went through the code and looks good 💯

**Comment by alamb** [comment]
@jorgecarleitao  -- @returnString did a great job explaining the rationale for schemas / catalogs from my perspective. 

Thanks for checking in on us and making sure it is clear what is going on

File: rust/datafusion/src/execution/context.rs
Comment (Dandandan) [reply rust/datafusion/src/execution/context.rs]: might be a good one for a `debug!` line actually.

**Comment by alamb** [comment]
Integration test failure seems unrelated

https://github.com/apache/arrow/pull/9818/checks?check_run_id=2219436309

**Pull request closed without merging**
